File: javascript/easy/binaryTreeTilt.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var findTilt = function(root) {
    return findTreeTilt(root).tilt;
};

var findTreeTilt = function(root) {
    if(root === null) return { tilt: 0, sum: 0 };
    const left = findTreeTilt(root.left);
    const right = findTreeTilt(root.right);
    const tilt = left.tilt + right.tilt + Math.abs(left.sum - right.sum);
    const sum = left.sum + right.sum + root.val;
    return { tilt: tilt, sum: sum };
};

File: ruby/easy/binary_tree_tilt.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Integer}
def find_tilt(root)
    obj = { tilt: 0 }
    find_tree_tilt(root, obj)
    return obj[:tilt]
end

def find_tree_tilt(root, obj)
    return 0 unless root
    left_sum = find_tree_tilt(root.left, obj)
    right_sum = find_tree_tilt(root.right, obj)
    obj[:tilt] += (left_sum - right_sum).abs
    return root.val + left_sum + right_sum
end
